### PR TITLE
Extremely WIP Hackney script

### PR DESCRIPTION
**Pull request by FreakyDug**
I think I've got this working about as well as it's going to given the state of the source.

A few known problems:
- There are a handful of rows in the original data that wrap onto a second line. There's probably a way to merge them back into where they should be but it didn't seem worth the trouble for 3 rows
- Because of the way they've generated the PDF from (I assume) a spreadsheet, in some cases cells overflow their column width. In these cases the overflow part gets put into the next column and is garbled. Not sure how to fix that, but again it's a small number affected, I think.

### CODE REVIEW

File: scripts/hackney.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    response = requests.get(START_URL)
+    soup = BeautifulSoup(response.text, "html.parser")
+    link = soup.find(class_="typepdf")["href"]
+    path = urlparse(link).path
+    drive_token = path.split("/")[3]
+    drive_url = DRIVE_DOWNLOAD_URL.format(drive_token)
+
+    dataframe = tabula.convert_into(drive_url, "hackney.csv", output_format="csv", pages="all")
+
+
```
Comment (jesspumphrey) [inline scripts/hackney.py:61]:
```suggestion
import csv
import pandas as pd

# Read the csv, using csv library to handle quotes
hackney_csv = open("hackney.csv", 'r')
reader = csv.reader(hackney_csv, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)
file_rows = []
for row in reader:
    file_rows.append(row)
    
# Merge together rows which should be merged
merged = []
# Holds the current row which we may need to merge subsequent rows into
current = file_rows[0].copy()

i = 1
while i < len(file_rows):
    row = file_rows[i]
    if len(row[1]) == 0:
        # If this column is blank, we need to merge this row into the current row
        for cell in range(len(row)):
            # Merge cell if it has any contents
            if len(row[cell]) > 0:
                current[cell] = current[cell] + ' ' + row[cell]
    else:
        # This should be a new row. Write the previous one and start a fresh current row
        merged.append(current)
        current = row.copy()
    i = i + 1
# Write the last row
raw_data.append(current)

# Remove redundant headers, and blank columns
clean_data = []
clean_data.append(raw_data[0])
for row in raw_data[1:]:
    # Skip headers beyond the first set
    if row[0].find('Licence No.') < 0:
        new_row = []
        for cell in range(len(row)):
            # These cells are blank, and cause the cells after them to become
            # misaligned with the headers, so remove them.
            if cell != 19 and cell != 21 and cell != 23:
                new_row.append(row[cell])
        clean_data.append(new_row)

df = pd.DataFrame(data=clean_data[1:], columns=clean_data[0])
print(df)
df.to_csv('out.csv')
```

**Review by jesspumphrey — COMMENTED** [review]
This seems to work for now, but it's pretty gnarly and will probably flake on a new PDF if it doesn't add the same blank columns.

**Comment by FreakyDug** [comment]
> 
> 
> This seems to work for now, but it's pretty gnarly and will probably flake on a new PDF if it doesn't add the same blank columns.

Thanks for this and sorry for my slow response, it's been a heavy week.

I tried running it locally and noticed that it's mixing the columns up like this:

```
,Licence No.,Property Address,Licence Holder Name,Licence Holder Address (business only)
0,,Star Estates & Lettings Ltd,,"400-402  West Green Road,  London,  N15 3PX."
1,,"1 Finsbury Park Road, London, N4 2LA",,End Terraced
2,,"1 Gunton Road, London, E5 9JT Mrs Nadia Kirwan",,Terraced
```

Number 2 has got `Property Address` and `Licence Holder Name` stuck together in one column. In the same place 0 has just licence holder name and 1 has just property address, because both of them are missing one of other column, but the data extracted from the PDF does not correctly idenitfy which column is missing.

Similarly, the last column in this cut down example should be `Licence Holder Address (business only)` but you can see in 1 and 2 it instead has got the value for `Property Type`.

I'm not sure there is a way to reliably get the values into the right columns because the data coming from tabula is combining multiple empty columns into one.

**Comment by JessP-utaw** [comment]
Can you separate them by the presence or absence of one of the other fields? Or a combination of them?

**Comment by FreakyDug** [comment]
> 
> 
> Can you separate them by the presence or absence of one of the other fields? Or a combination of them?

Hmm, it does look like here's some correlation between license type and which fields are missing 🤔

**Comment by FreakyDug** [comment]
I discovered that you can pass Tabula a list of column positions to override it's guesses at where each other begins. That seems to have resolved the merged columns problem. 🎉

File: scripts/hackney.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+from urllib.parse import urlparse
+
+from bs4 import BeautifulSoup
+import pandas
+import requests
+import tabula
+
+START_URL = "https://hackney.gov.uk/hmo"
+DRIVE_DOWNLOAD_URL = "https://drive.google.com/uc?export=download&id={}"
```
Comment (chrisdevereux) [inline scripts/hackney.py:16]: Is the output data being downloaded to google drive? Doesn't need to block merging, but maybe agreeing on where this data gets sent would be a good idea. I was wondering if we could use github actions to run this on a schedule and just commit the csvs to master

Comment (FreakyDug) [reply scripts/hackney.py:16]: Oh, no, the opposite actually. Hackney are hosting their PDF of the data on Google drive, so I'm downloading it from there and then spitting it out as a CSV and committing that to the repo.

Comment (chrisdevereux) [reply scripts/hackney.py:16]: ah! makes sense

**Comment by chrisdevereux** [comment]
Looks good to me!

**Review by chrisdevereux — APPROVED** [review]

**Pull request merged**